engine/src/main/java/org/hibernate/search/engine/backend/document/DocumentElement.java

```diff
@@ -16,4 +16,34 @@
  */
 public interface DocumentElement {
 
+	/**
+	 * Add a new value to the referenced field in this document element.
+	 * <p>
+	 * This method can be called multiple times for the same field,
+	 * which will result in multiple values being added to the same field.
+	 *
+	 * @param fieldReference The field to add a value to.
+	 * @param value The value to add to the field.
+	 */
+	<F> void add(IndexFieldReference<F> fieldReference, F value);
+
+	/**
+	 * Add a new child object to the referenced field in this document element.
+	 *
+	 * @param fieldReference The object field to add a child to.
+	 * @return The new child object, that can be populated with its own fields.
+	 */
+	DocumentElement addChild(IndexObjectFieldReference fieldReference);
```

Comment:
I agree it's not perfect. `addDocument` seems confusing, because:

1. Technically, we're not always adding a nested document. Depending on the storage type of the object field, this "child" may be completely virtual, storing its content in its parent.
2. In the Elasticsearch documentation, they do not talk about "documents" event when they are implemented as nested documents. They talk about "object" properties or "nested" properties, but "document" always refer to the root document.
3. In Hibernate Search, we call these "object fields", "objects", "document elements", but never actual documents, because they may not be documents.

I think I could go with `addObject`, if you agree. Which would be consistent with `IndexObjectFieldReference`. In essence a "document element" is either a "document" (the root) or an "object" (the non-root nodes).